cocoa/src/toga_cocoa/window.py

```diff
@@ -288,11 +355,127 @@ def get_visible(self):
     # Window state
     ######################################################################
 
-    def set_full_screen(self, is_full_screen):
-        current_state = bool(self.native.styleMask & NSWindowStyleMask.FullScreen)
-        if is_full_screen != current_state:
+    def get_window_state(self, actual_state=True):
+        if not actual_state and self._pending_state_transition:
+            return self._pending_state_transition
+        if bool(self.container.native.isInFullScreenMode()):
+            return WindowState.PRESENTATION
+        elif bool(self.native.styleMask & NSWindowStyleMask.FullScreen):
+            return WindowState.FULLSCREEN
+        elif bool(self.native.isZoomed):
+            return WindowState.MAXIMIZED
+        elif bool(self.native.isMiniaturized):
+            return WindowState.MINIMIZED
+        else:
+            return WindowState.NORMAL
+
+    def set_window_state(self, state):
+        # Since the requests to the OS for changing window states are non-blocking,
+        # if we are in the middle of processing a state, we need to store the
+        # user-requested state and apply the state when we have completed processing
+        # a transition. There are 2 types of callbacks:
+        # * EnteredState:
+        #   Here, we need to check if the current state is the same as the pending
+        #   state.
+        #   -- If yes: Clear the pending state variable and return.
+        #   -- If no: Apply NORMAL state, which will later apply the pending state
+        #             when the state is NORMAL.
+        # * ExitedState:
+        #   Here, since we are in NORMAL state, we just apply the pending state.
+        #   When we enter the user-requested pending state, then clear the pending
+        #   state variable and return.
+
+        if self._pending_state_transition:
+            self._pending_state_transition = state
+        else:
+            # If the app is in presentation mode, but this window isn't, then
+            # exit app presentation mode before setting the requested state.
+            if any(
+                window.state == WindowState.PRESENTATION and window != self.interface
+                for window in self.interface.app.windows
+            ):
+                self.interface.app.exit_presentation_mode()
+
+            self._pending_state_transition = state
+            if self.get_window_state() != WindowState.NORMAL:
+                self._apply_state(WindowState.NORMAL)
+            else:
+                self._apply_state(state)
+
+    def _apply_state(self, target_state):
+        current_state = self.get_window_state()
+        if target_state is None:
+            return
+
+        elif target_state == current_state:
+            self._pending_state_transition = None
+            return
```

Comment:
>Consider the case where I'm in NORMAL, and I've requested FULLSCREEN, but that transition hasn't completed.

Here, the methods called and attributes set will be:
```
set_window_state(FULLSCREEN)
self._pending_state_transition = FULLSCREEN
self._apply_state(FULLSCREEN)
```
>I then request a transition to NORMAL

```
set_window_state(NORMAL)
```
Since, `self._pending_state_transition` would be set, so it will do:
```
self._pending_state_transition = NORMAL
```
and then just return, without actually applying NORMAL. Then when the FULLSCREEN transition is complete, it will apply the new pending state.

>I'm actually not convinced this clause is needed at all. The interface-level checks whether there's a pending transition to the requested state; the role of this method is to move the window closer to the desired "pending" state (which is whatever was requested by the user)

The reason, why it is required on non-blocking API backends is because of processing of non-blocking APIs. For example:
https://github.com/beeware/toga/blob/0d6732283b3ff0ffb5473cfceadfab9ef8c0ad48/cocoa/src/toga_cocoa/window.py#L70-L80

As we see here, on the completion of the non blocking API(`setIsMiniaturized(True)`), `_apply_state()` is again called, and the same state check on `_apply_state()` is used to terminate further processing.
https://github.com/beeware/toga/blob/0d6732283b3ff0ffb5473cfceadfab9ef8c0ad48/cocoa/src/toga_cocoa/window.py#L405-L412
Hence, the check is required on backends with non-blocking API. However, the check is not required on backends with blocking API like WinForms and Android. So, I have removed it from those backends.

Reply:
> > Consider the case where I'm in NORMAL, and I've requested FULLSCREEN, but that transition hasn't completed.
...
> Since, `self._pending_state_transition` would be set, so it will do:
> 
> ```
> self._pending_state_transition = NORMAL
> ```
> 
> and then just return, without actually applying NORMAL. Then when the FULLSCREEN transition is complete, it will apply the new pending state.

Ok - I wasn't factoring in the impact of the pre-processing step in `set_window_state()`.